internal/graph/check_fast_path.go

```diff
@@ -0,0 +1,674 @@
+			// we need to ensure we have all iterators at all times
+			continue
+		}
+
+		// all children have the same value
+		if len(indexes) == childrenTotal {
+			// all entries are the same thus flush entry and move iterators
+			// there should only be 1 value
+			for idx, iterIdx := range indexes {
+				t, err := iterProducers[iterIdx].iter.Next(ctx)
+				if err != nil {
+					if storage.IterIsDoneOrCancelled(err) {
+						iterProducers[iterIdx].iter.Stop()
+						iterProducers[iterIdx].iter = nil
+						break
+					}
+					concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+					return
+				}
+				if idx == 0 {
+					batch = append(batch, t)
+				}
+			}
+
+			if len(batch) > IteratorMinBatchThreshold {
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{iter: storage.NewStaticTupleKeyIterator(batch)}, iterChan)
+				batch = make([]*openfgav1.TupleKey, 0)
+			}
+			continue
+		}
+
+		// move all entries to less than the MAX to be >= than MAX
+		for _, producer := range iterProducers {
+			t, err := producer.iter.Head(ctx)
+			if err != nil {
+				if storage.IterIsDoneOrCancelled(err) {
+					// this is highly unlikely due to the previous check
+					producer.iter.Stop()
+					producer.iter = nil
+					continue
+				}
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+				return
+			}
+			tmpKey := t.GetObject()
+			for tmpKey < maxKey {
+				t, err := producer.iter.Next(ctx)
+				if err != nil {
+					if storage.IterIsDoneOrCancelled(err) {
+						producer.iter.Stop()
+						producer.iter = nil
+						break
+					}
+					concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+					return
+				}
+				tmpKey = t.GetObject()
+			}
+		}
+	}
+}
+
+func fastPathDifference(ctx context.Context, iterProducers []*iteratorProducer, iterChan chan *iteratorMsg) {
+	batch := make([]*openfgav1.TupleKey, 0)
+
+	defer func() {
+		// flush
+		if len(batch) > 0 {
+			concurrency.TrySendThroughChannel(ctx, &iteratorMsg{iter: storage.NewStaticTupleKeyIterator(batch)}, iterChan)
+		}
+		close(iterChan)
+		cleanupIteratorProducers(iterProducers)
+	}()
+
+	// both base and difference are still remaining
+	for len(iterProducers) == 2 {
+		if ctx.Err() != nil {
+			return
+		}
+		var err error
+		iterProducers, err = pollIteratorProducers(ctx, iterProducers)
+		if err != nil {
+			concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+			return
+		}
+		if len(iterProducers) != 2 {
+			// short circuit
+			break
+		}
+
+		allIters := true
+		base := ""
+		diff := ""
+		for idx, producer := range iterProducers {
+			v, err := producer.iter.Head(ctx)
+			if err != nil {
+				if storage.IterIsDoneOrCancelled(err) {
+					producer.iter.Stop()
+					producer.iter = nil
+					allIters = false
+					// we need to ensure we have all iterators at all times
+					break
+				}
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+				return
+			}
+			if idx == 0 {
```

Comment:
Nit / suggestion:

```go
const base = 0
const diff = 1
```

and then use these indexes throughout the function. It could make it more readable.